<a href="https://colab.research.google.com/github/ANANYAsindhe/Data_Analysis_Project/blob/main/Business_Analytics_with_AI_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

# Load the CSV file into a DataFrame
file_path = '/content/employees_dataset.xlsx'
df = pd.read_excel('/content/employees_dataset.xlsx')

# Display the first few rows to confirm it loaded correctly
print(df.head())

   employee_id         department     region  education gender  \
0         8724         Technology  region_26  Bachelors      m   
1        74430                 HR   region_4  Bachelors      f   
2        72255  Sales & Marketing  region_13  Bachelors      m   
3        38562        Procurement   region_2  Bachelors      f   
4        64486            Finance  region_29  Bachelors      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   24                   NaN   
1               other                1   31                   3.0   
2               other                1   31                   1.0   
3               other                3   31                   2.0   
4            sourcing                1   30                   4.0   

   length_of_service  KPIs_met_more_than_80  awards_won  avg_training_score  
0                  1                      1           0                  77  
1                  5            

In [3]:
# Check for missing values in each column
print("Missing values per column:")
print(df.isnull().sum())

# Check for duplicate records
print("\nNumber of duplicate records:")
print(df.duplicated().sum())

Missing values per column:
employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

Number of duplicate records:
2


# Task
Analyze employee performance and retention patterns using the dataset "employees_dataset.xlsx - employees_dataset.csv" and provide actionable insights.

## Handle missing values

### Subtask:
Inspect and decide on a strategy for handling missing values in the 'education' and 'previous_year_year_rating' columns. This could involve imputation or removal depending on the analysis requirements.


**Reasoning**:
Inspect the number and distribution of missing values in 'education' and 'previous_year_rating' to inform the handling strategy.



In [4]:
# Inspect missing values in 'education'
print("Missing values in 'education':", df['education'].isnull().sum())

# Inspect the distribution of 'education'
print("\nDistribution of 'education':")
print(df['education'].value_counts(dropna=False))

# Inspect missing values in 'previous_year_rating'
print("\nMissing values in 'previous_year_rating':", df['previous_year_rating'].isnull().sum())

# Inspect the distribution of 'previous_year_rating'
print("\nDistribution of 'previous_year_rating':")
print(df['previous_year_rating'].value_counts(dropna=False))

Missing values in 'education': 771

Distribution of 'education':
education
Bachelors          11519
Masters & above     4841
NaN                  771
Below Secondary      286
Name: count, dtype: int64

Missing values in 'previous_year_rating': 1363

Distribution of 'previous_year_rating':
previous_year_rating
3.0    5813
5.0    3817
4.0    3156
1.0    1976
NaN    1363
2.0    1292
Name: count, dtype: int64


**Reasoning**:
Based on the distributions, imputing 'education' with the mode ('Bachelors') and 'previous_year_rating' with a constant value (e.g., 0 to indicate no rating) seems reasonable given the categorical nature of 'education' and the ordinal nature of 'previous_year_rating' where missing values could imply no prior rating.



In [5]:
# Impute missing 'education' values with the mode
mode_education = df['education'].mode()[0]
df['education'].fillna(mode_education, inplace=True)

# Impute missing 'previous_year_rating' values with 0
df['previous_year_rating'].fillna(0, inplace=True)

# Verify that missing values have been handled
print("\nMissing values after imputation:")
print(df[['education', 'previous_year_rating']].isnull().sum())


Missing values after imputation:
education               0
previous_year_rating    0
dtype: int64


/tmp/ipython-input-2942767500.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['education'].fillna(mode_education, inplace=True)
/tmp/ipython-input-2942767500.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', tr

## Handle duplicate records

### Subtask:
Inspect and remove or handle the duplicate records identified in the dataset.


**Reasoning**:
Inspect the duplicate rows in the DataFrame to understand their nature by filtering the DataFrame to show only the duplicated rows.



In [6]:
# Inspect the duplicate rows
duplicate_rows = df[df.duplicated(keep=False)]
print("Duplicate rows:")
display(duplicate_rows)

Duplicate rows:


employee_id department    region  education gender recruitment_channel  \
17413        49584         HR  region_7  Bachelors      m               other   
17414        49584         HR  region_7  Bachelors      m               other   
17415        49584         HR  region_7  Bachelors      m               other   

       no_of_trainings  age  previous_year_rating  length_of_service  \
17413                1   33                   1.0                  9   
17414                1   33                   1.0                  9   
17415                1   33                   1.0                  9   

       KPIs_met_more_than_80  awards_won  avg_training_score  
17413                      0           0                  51  
17414                      0           0                  51  
17415                      0           0                  51

**Reasoning**:
Remove the identified duplicate rows from the DataFrame and verify that the duplicate rows have been successfully removed by checking the number of duplicate records again.



In [7]:
# Remove duplicate rows, keeping the first occurrence
df.drop_duplicates(inplace=True)

# Verify that the duplicate rows have been removed
print("\nNumber of duplicate records after removal:")
print(df.duplicated().sum())


Number of duplicate records after removal:
0


## Data transformation

### Subtask:
Perform any necessary data type conversions or feature engineering required for the analysis.


**Reasoning**:
Convert 'previous_year_rating' to integer, create 'has_awards' feature, and create 'age_group' feature by binning 'age'.



In [8]:
# 1. Convert 'previous_year_rating' to integer
df['previous_year_rating'] = df['previous_year_rating'].astype(int)

# 2. Create 'has_awards' binary feature
df['has_awards'] = (df['awards_won'] > 0).astype(int)

# 3. Create 'age_group' categorical feature by binning 'age'
bins = [20, 30, 40, 50, 60, 70]
labels = ['20-30', '31-40', '41-50', '51-60', '61-70']
df['age_group'] = pd.cut(df['age'], bins=bins, labels=labels, right=True, include_lowest=True)

# Display the first few rows with the new features and updated type
print(df[['previous_year_rating', 'has_awards', 'age_group', 'age']].head())

# Verify the data types of the new and converted columns
print("\nData types after conversion and feature engineering:")
print(df[['previous_year_rating', 'has_awards', 'age_group']].dtypes)

   previous_year_rating  has_awards age_group  age
0                     0           0     20-30   24
1                     3           0     31-40   31
2                     1           0     31-40   31
3                     2           0     31-40   31
4                     4           0     20-30   30

Data types after conversion and feature engineering:
previous_year_rating       int64
has_awards                 int64
age_group               category
dtype: object


## Identify performance patterns

### Subtask:
Analyze how factors like age, education, and department relate to employee performance.


**Reasoning**:
Calculate and print the average training score for each age group, education level, and department, and then calculate and print the proportion of employees meeting KPIs for each of these categories.



In [9]:
# Calculate the average avg_training_score for each age_group
avg_training_score_by_age_group = df.groupby('age_group')['avg_training_score'].mean()
print("Average Training Score by Age Group:")
print(avg_training_score_by_age_group)

# Calculate the average avg_training_score for each education level
avg_training_score_by_education = df.groupby('education')['avg_training_score'].mean()
print("\nAverage Training Score by Education Level:")
print(avg_training_score_by_education)

# Calculate the average avg_training_score for each department
avg_training_score_by_department = df.groupby('department')['avg_training_score'].mean()
print("\nAverage Training Score by Department:")
print(avg_training_score_by_department)

# Calculate the proportion of employees who met KPIs for each age_group
kpi_proportion_by_age_group = df.groupby('age_group')['KPIs_met_more_than_80'].mean()
print("\nProportion of Employees Meeting KPIs by Age Group:")
print(kpi_proportion_by_age_group)

# Calculate the proportion of employees who met KPIs for each education level
kpi_proportion_by_education = df.groupby('education')['KPIs_met_more_than_80'].mean()
print("\nProportion of Employees Meeting KPIs by Education Level:")
print(kpi_proportion_by_education)

# Calculate the proportion of employees who met KPIs for each department
kpi_proportion_by_department = df.groupby('department')['KPIs_met_more_than_80'].mean()
print("\nProportion of Employees Meeting KPIs by Department:")
print(kpi_proportion_by_department)

Average Training Score by Age Group:
age_group
20-30    63.746890
31-40    63.155175
41-50    61.892078
51-60    63.254839
61-70          NaN
Name: avg_training_score, dtype: float64

Average Training Score by Education Level:
education
Bachelors          62.737467
Below Secondary    65.902098
Masters & above    64.134270
Name: avg_training_score, dtype: float64

Average Training Score by Department:
department
Analytics            84.565115
Finance              60.326683
HR                   50.386282
Legal                59.533133
Operations           60.351589
Procurement          70.180357
R&D                  84.451807
Sales & Marketing    50.056614
Technology           79.847658
Name: avg_training_score, dtype: float64

Proportion of Employees Meeting KPIs by Age Group:
age_group
20-30    0.363856
31-40    0.368357
41-50    0.331447
51-60    0.317204
61-70         NaN
Name: KPIs_met_more_than_80, dtype: float64

Proportion of Employees Meeting KPIs by Education Level:
education
B

/tmp/ipython-input-793722522.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  avg_training_score_by_age_group = df.groupby('age_group')['avg_training_score'].mean()
/tmp/ipython-input-793722522.py:17: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  kpi_proportion_by_age_group = df.groupby('age_group')['KPIs_met_more_than_80'].mean()


## Examine retention trends

### Subtask:
Look at how long employees stay (tenure) and how this is influenced by their performance ratings and awards won.


**Reasoning**:
Calculate the average length of service for employees based on their previous year's rating and whether they won awards, then print the results to understand the relationship between these factors and tenure.



In [10]:
# Calculate the average length of service for employees based on their 'previous_year_rating'
avg_tenure_by_rating = df.groupby('previous_year_rating')['length_of_service'].mean()
print("Average Length of Service by Previous Year Rating:")
print(avg_tenure_by_rating)

# Calculate the average length of service for employees based on whether they received 'awards_won'
avg_tenure_by_awards = df.groupby('has_awards')['length_of_service'].mean()
print("\nAverage Length of Service by Awards Won (0: No Awards, 1: Awards Won):")
print(avg_tenure_by_awards)

Average Length of Service by Previous Year Rating:
previous_year_rating
0    1.000000
1    6.199595
2    6.075851
3    6.200413
4    6.218631
5    6.264868
Name: length_of_service, dtype: float64

Average Length of Service by Awards Won (0: No Awards, 1: Awards Won):
has_awards
0    5.826846
1    4.742015
Name: length_of_service, dtype: float64


## Provide recommendations

### Subtask:
Based on the analysis, provide recommendations to HR managers on how to improve employee engagement, productivity, and reduce turnover.


**Reasoning**:
Synthesize the findings from the previous analysis steps and formulate actionable recommendations for HR managers based on these insights to improve employee engagement, productivity, and reduce turnover.



In [11]:
# Based on "Identify performance patterns" findings:
# - Age Group and Performance: 20-40 age group performs slightly better.
# - Education and Performance: "Below Secondary" and "Masters & above" perform well in KPIs.
# - Department and Performance: Analytics, R&D, and Technology excel in training scores; Operations and R&D in KPIs; Sales & Marketing lowest in KPIs.

# Based on "Examine retention trends" findings:
# - Performance Rating and Tenure: Employees with any rating (1-5) have longer tenure than those with no rating (0).
# - Awards and Tenure: Counter-intuitively, employees without awards have slightly longer average tenure.

print("Recommendations for HR Managers:")
print("\n1. Targeted Training Programs:")
print("- Develop specialized training programs for departments with lower average training scores and KPI achievement, such as Sales & Marketing.")
print("- Investigate the training needs of the 41-60 age group to understand potential reasons for slightly lower average performance compared to younger groups.")
print("- Analyze the factors contributing to the high performance of employees with 'Below Secondary' and 'Masters & above' education to potentially replicate successful strategies.")

print("\n2. Enhanced Performance Management:")
print("- Implement a clear performance review process for all employees, ensuring everyone receives a 'previous_year_rating'. Focus on providing timely feedback and development plans, especially for new hires (those currently with a rating of 0) to potentially increase retention.")
print("- Investigate the surprising finding regarding awards and tenure. Understand if employees receiving awards are being promoted or transitioning out for other reasons. Re-evaluate the recognition program to ensure it positively impacts retention.")

print("\n3. Improved Employee Engagement and Support:")
print("- Explore reasons for lower KPI achievement in departments like Sales & Marketing and implement targeted support or resources.")
print("- Consider mentorship programs pairing high-performing employees (potentially from R&D, Operations, Analytics) with those in departments or age groups needing improvement.")
print("- Further investigate the retention differences based on previous year's rating (0 vs. 1-5) to identify potential early disengagement signs and implement interventions.")

Recommendations for HR Managers:

1. Targeted Training Programs:
- Develop specialized training programs for departments with lower average training scores and KPI achievement, such as Sales & Marketing.
- Investigate the training needs of the 41-60 age group to understand potential reasons for slightly lower average performance compared to younger groups.
- Analyze the factors contributing to the high performance of employees with 'Below Secondary' and 'Masters & above' education to potentially replicate successful strategies.

2. Enhanced Performance Management:
- Implement a clear performance review process for all employees, ensuring everyone receives a 'previous_year_rating'. Focus on providing timely feedback and development plans, especially for new hires (those currently with a rating of 0) to potentially increase retention.
- Investigate the surprising finding regarding awards and tenure. Understand if employees receiving awards are being promoted or transitioning out for othe

## Summary:

### Data Analysis Key Findings

*   There were 771 missing values in the 'education' column and 1363 missing values in the 'previous_year_rating' column.
*   Duplicate records were identified and removed from the dataset.
*   The 'previous_year_rating' column was converted to an integer data type.
*   A binary feature 'has\_awards' was created, indicating whether an employee won awards.
*   A categorical feature 'age\_group' was created by binning the 'age' column.
*   The average training scores and the proportion of employees meeting KPIs varied across age groups, education levels, and departments.
*   Employees with 'Below Secondary' education had the highest average training scores and the highest proportion of meeting KPIs.
*   Departments like Analytics, R\&D, and Technology had notably higher average training scores, while Operations and R\&D showed higher proportions of employees meeting KPIs.
*   Employees with any previous year's rating (1-5) had a significantly longer average length of service (around 6.1-6.2 years) compared to those with a rating of 0 (1 year).
*   Employees who did not win awards had a slightly higher average length of service (5.83 years) compared to those who did win awards (4.74 years).

### Insights or Next Steps

*   Implement targeted training programs for departments and age groups with lower performance metrics to improve overall employee productivity.
*   Investigate the reasons behind the shorter tenure of employees with no previous year's rating and the counter-intuitive finding regarding awards and tenure to refine retention strategies and recognition programs.
